### PR TITLE
[TIMOB-23422] Try and fix HTTPClient post with Blob and string data

### PR DESCRIPTION
**Pull request by sgtcoolguy**
**JIRA**: https://jira.appcelerator.org/browse/TIMOB-23422

Looks like we're mangling the post data we send as multi-part form data. I'm not sure we can properly convert everything to `std::vector<std::uint8_t>` like we currently do. So this PR goes back to keep the values as JSValue and letting the Windows implementation look for Blobs and when found basically just copy their data as a buffer passthrough, otherwise cast to std::string and convert to Platform::String as normal string content.

I created a new requestbin used below, and tested with the following code. I still need to hit it with iOS/Android to compare the exact data from the PNG file matches, but the request looks closer. (Note that Android actually writes a blob as a file to disk and then uses Apache's FileBody to attach it - whereas we pass the bytes in a buffer without re-writing to disk).

http://requestb.in/1j1lh941?inspect

```javascript
var url = 'http://requestb.in/1j1lh941';
//use favorite inspector here...

function go() {
    var imageFile = Ti.Filesystem.getFile(Ti.Filesystem.resourcesDirectory, "Logo.png");
    var newId = new Date().getTime();
    var newName = 'HEY_YOU_GUYS_WAIT_FOR_ME-' + newId;
    var client = Ti.Network.createHTTPClient({
        onload: function (e) {
            alert('success');
        },
        onerror: function (err) {
            alert('error');
        }
    });

    client.open("POST", url);

    var f = {
        name: newName,
        attachment: imageFile.read(imageFile)
    };

    client.send(f);
      
}


Titanium.UI.setBackgroundColor('#fff');

var win = Titanium.UI.createWindow({
    title: 'win',
    backgroundColor: '#fff'
});

var label = Ti.UI.createLabel({
    color: 'blue',
    text: 'Send it!',
    width: 300, height: 200
});

label.addEventListener('click', function (e) {
    go();
});

win.add(label);

win.open();
```

### CODE REVIEW

**Comment by sgtcoolguy** [comment]
So, after testing this with iPhone and Android, this looks correct to me. We match the same content body for the same PNG as Android now. Oddly enough, iPhone doesn't seem to be picking up the PNG properly, maybe a parity issue in the file path?

I can say that the requests are _not_ identical, even excluding headers. 

Heres what **Windows** sends with this PR:
```
RAW BODY

--7158ebc6-df3a-4079-ac8c-b19f391f5ec4
Content-Length: 801
Content-Disposition: form-data; name="attachment"

PNG


IHDR<qâtEXtSoftwareAdobe ImageReadyqÉe<ÃIDATxÚìÝ½M#Aa¼"tABv]@~k ¢®ËÉh(ÃùÞén|3»;ß÷¼ÒÊ-Ùy4³¶ü³ÇñBªÝ`ÀXXK`I`	,%%°ÀXXK`I`	,%%°ÀXXK`I`	,%%°ºËë«éúèÆ%[¡u+üª}9î7Í{¤IÇñª\<ckÅZ¦ÃÀ?'*°ª¬TápEAÕ3¬ûh¸&P=5SÇsª0¸¦Pçú\PE8yïWTT^v+2ª°zÄUX=áÊ*¬peAÖqûÞfAÖqeCÖpýê*ªÐ°Ö++ªð°ÄU
XKàÊ*
¬9qM úU*Xsà@õRîû9ÓX§ûÌ{+\'P=dç_¦¨*°ªã
¬ê¸ «:.¨Àª*°ªã
¬ê¸ «	®rì «®k¨Àj*°ªázûÇÕ¿XgUNÖw_lú°æEõcâ&[¸Àª*Ôµ"Tåë
.°j£Úx+.°ÎCuâ­¸À:\`5CXÍPÁV3TpÕ\`}¢º«
®ä°Êä@Ý¶@WRXGT»¨àJkNTp%µ*¸ÃZUv\C2Tû9QeÆ5@XÿjÑbÍk
.°:G	×\`A×\`C×Õúq5_?#¡ú.°fj
Õ	\Ýu`.ÈÊöx!+ÀÎØÒm!²b	,%%°ÀXXK`I`	,%%°ÀXXK`I`	,%%°ÀXXK`I`	,%%°ÀXXZ²?Ê_t©&wñRIEND®B`
--7158ebc6-df3a-4079-ac8c-b19f391f5ec4
Content-Length: 38
Content-Type: text/plain; charset=UTF-8
Content-Disposition: form-data; name="name"

HEY_YOU_GUYS_WAIT_FOR_ME-1464030733076
--7158ebc6-df3a-4079-ac8c-b19f391f5ec4--
```

and **Android**:
```
RAW BODY

--_VleL1V6AYwezhi2tCAUl6dShGs_YLhxyB-jVi
Content-Disposition: form-data; name="name"
Content-Transfer-Encoding: 8bit

HEY_YOU_GUYS_WAIT_FOR_ME-1464032905507
--_VleL1V6AYwezhi2tCAUl6dShGs_YLhxyB-jVi
Content-Disposition: form-data; name="attachment"; filename="tixhr1862771948.png"
Content-Type: image/png
Content-Transfer-Encoding: binary

PNG


IHDR<qâtEXtSoftwareAdobe ImageReadyqÉe<ÃIDATxÚìÝ½M#Aa¼"tABv]@~k ¢®ËÉh(ÃùÞén|3»;ß÷¼ÒÊ-Ùy4³¶ü³ÇñBªÝ`ÀXXK`I`	,%%°ÀXXK`I`	,%%°ÀXXK`I`	,%%°ºËë«éúèÆ%[¡u+üª}9î7Í{¤IÇñª\<ckÅZ¦ÃÀ?'*°ª¬TápEAÕ3¬ûh¸&P=5SÇsª0¸¦Pçú\PE8yïWTT^v+2ª°zÄUX=áÊ*¬peAÖqûÞfAÖqeCÖpýê*ªÐ°Ö++ªð°ÄU
XKàÊ*
¬9qM úU*Xsà@õRîû9ÓX§ûÌ{+\'P=dç_¦¨*°ªã
¬ê¸ «:.¨Àª*°ªã
¬ê¸ «	®rì «®k¨Àj*°ªázûÇÕ¿XgUNÖw_lú°æEõcâ&[¸Àª*Ôµ"Tåë
.°j£Úx+.°ÎCuâ­¸À:\`5CXÍPÁV3TpÕ\`}¢º«
®ä°Êä@Ý¶@WRXGT»¨àJkNTp%µ*¸ÃZUv\C2Tû9QeÆ5@XÿjÑbÍk
.°:G	×\`A×\`C×Õúq5_?#¡ú.°fj
Õ	\Ýu`.ÈÊöx!+ÀÎØÒm!²b	,%%°ÀXXK`I`	,%%°ÀXXK`I`	,%%°ÀXXK`I`	,%%°ÀXXZ²?Ê_t©&wñRIEND®B`
--_VleL1V6AYwezhi2tCAUl6dShGs_YLhxyB-jVi--
```

File: Source/Network/src/HTTPClient.cpp [hunk 1]
```diff
@@ -371,26 +378,13 @@ namespace TitaniumWindows
 			}
 		}
 
-		Windows::Storage::Streams::Buffer^ HTTPClient::charVecToBuffer(std::vector<std::uint8_t> char_vector)
+		Windows::Storage::Streams::IBuffer^ HTTPClient::charVecToBuffer(std::vector<std::uint8_t> char_vector)
```
Comment (sgtcoolguy) [inline Source/Network/src/HTTPClient.cpp]: I stole this code from File, maybe it should be moved into TitaniumWindows::Utility?

**Comment by infosia** [comment]
Looks good. Can we have unit tests for this?

**Comment by sgtcoolguy** [comment]
Yeah, I was wondering that as well. It'd be ugly to try and generate a request.in URL each time and parse the page results... I'm curious if there's some other test server setup that would echo what we send it or something so we can test that way?

**Comment by garymathews** [comment]
httpbin?

```Javascript
var win = Titanium.UI.createWindow(),
    btn = Ti.UI.createButton({backgroundColor: 'gray', title: 'HTTP POST'});

btn.addEventListener('click', function (e) {
    var imageFile = Ti.Filesystem.getFile(Ti.Filesystem.resourcesDirectory, "Logo.png"),
        name = 'HEY_YOU_GUYS_WAIT_FOR_ME-' + new Date().getTime(),
        client = Ti.Network.createHTTPClient({
            onload: function (e) {
                Ti.API.info(this.responseText);
            },
            onerror: function (e) {
                alert(e.error);
            }
        });
    client.open("POST", 'http://httpbin.org/post');
    client.send({name: name, attachment: imageFile.read(imageFile)});
});

win.add(btn);
win.open();
```

**Comment by sgtcoolguy** [comment]
Ok, just pushed a unit test for doing a POST of multipart/form-data with a Ti.Blob member against httpbin.org

File: Examples/NMocha/src/Assets/ti.network.httpclient.test.js [hunk 2]
```diff
@@ -367,4 +376,56 @@ describe("Titanium.Network.HTTPClient", function () {
...
+                should(result.headers).have.property('Content-Type');
+                should(result.headers['Content-Type']).startWith('multipart/form-data');
+
+                // check name got added
+                should(result).have.property('form');
+                should(result.form).have.property('name');
+                should(result.form.name).eql(newName);
+
+                // check blob data
+                should(result).have.property('files');
```
Comment (garymathews) [inline Examples/NMocha/src/Assets/ti.network.httpclient.test.js]:
```Javascript
should(result).have.property('form');
should(result.form).have.property('attachment');
```

I don't know how we can validate the ``attachment``, it's interpreted as a string and not binary data

**Comment by sgtcoolguy** [comment]
I see the test failed on Windows 10 desktop. Not sure why, it does run and pass locally for me now.

**Comment by sgtcoolguy** [comment]
Kind of off-topic, but it's incredibly annoying that assertions failing inside JS callbacks effectively get squashed in our unit tests. I can't easily tell what actually failed in these unit tests, because the failure in unload is getting "squashed" and instead we end up timing out the test. @infosia any ideas on what we could do to surface runtime Errors that happen in JS callbacks - so that it gets reported up to ti-mocha and we get the real failure reason?

**Comment by sgtcoolguy** [comment]
https://jira.appcelerator.org/browse/TIMOB-23439

**Comment by garymathews** [comment]
FR: **PASS**

Local Mocha test also passes for me too

**Pull request merged**